firedrake/slate/slac/utils.py

```diff
@@ -137,6 +138,47 @@ def visit_Symbol(self, o, *args, **kwargs):
         return SymbolWithFuncallIndexing(o.symbol, o.rank, o.offset)
 
 
+def topological_sort(exprs):
+    """Returns a topologically sorted list of Slate expressions.
+    The list is generated by examining the dependencies of each
+    expression and moving nodes that have all its dependencies
+    resolved. A node is resolved once all the nodes in its dependencies
+    have been moved into the sorted list.
+
+    A list of expressions, say [D, C, B, A], is provided and a dependency
+    graph is generated based on the operands of each expression. If
+    the dependency graph looks like
+
+    G = { D: [B, A],
+          C: [B],
+          B: [],
+          A: [] }
+
+    i.e., D depends on A and B, C depends on B, and A and B don't depend
+    on any expression, then [A, B, C, D] is returned.
+
+    :arg exprs: An iterable of Slate nodes.
+    """
+    # Generate the graph as a dictionary whose keys are the Slate nodes.
+    # Don't include the expr itself in its dependencies
+    dependency_info = [(expr, set(traverse_dags([expr])) - {expr})
+                       for expr in exprs]
+    dependency_graph = OrderedDict(dependency_info)
+
+    schedule = []
+    while dependency_graph:
+        for expr in list(dependency_graph.keys()):
+            if not dependency_graph[expr].intersection(dependency_graph):
+                schedule.append(expr)
+                break
+        else:
+            raise RuntimeError("A cyclic dependency is found.")
+
+        del dependency_graph[expr]
```

Comment:
This algorithm is, I guess, no worse than the original here, but _come on_: DFS is also simple and much faster.